internal/controller/operator/factory/vmagent/vmagent.go

```diff
@@ -146,125 +142,38 @@ func CreateOrUpdate(ctx context.Context, cr *vmv1beta1.VMAgent, rclient client.C
-		if err != nil {
-			return fmt.Errorf("cannot build new deploy for vmagent: %w", err)
-		}
-	}
-	newDeploy, err := newDeploy(cr, ac)
+	newAppTpl, err := newK8sApp(cr, ac)
 	if err != nil {
 		return fmt.Errorf("cannot build new deploy for vmagent: %w", err)
 	}
-
-	if !cr.Spec.DaemonSetMode && cr.Spec.ShardCount != nil && *cr.Spec.ShardCount > 1 {
-		return createOrUpdateShardedDeploy(ctx, rclient, cr, prevCR, newDeploy, prevDeploy)
-	}
-	return createOrUpdateDeploy(ctx, rclient, cr, prevCR, newDeploy, prevDeploy)
-}
-
-func createOrUpdateDeploy(ctx context.Context, rclient client.Client, cr, _ *vmv1beta1.VMAgent, newDeploy, prevObjectSpec runtime.Object) error {
-	deploymentNames := make(map[string]struct{})
-	stsNames := make(map[string]struct{})
-
-	var err error
-	switch newDeploy := newDeploy.(type) {
-	case *appsv1.Deployment:
-		var prevDeploy *appsv1.Deployment
-		if prevObjectSpec != nil {
-			prevAppObject, ok := prevObjectSpec.(*appsv1.Deployment)
-			if ok {
-				prevDeploy = prevAppObject
-				prevDeploy, err = k8stools.RenderPlaceholders(prevDeploy, defaultPlaceholders)
-				if err != nil {
-					return fmt.Errorf("cannot fill placeholders for prev deployment in vmagent: %w", err)
-				}
-			}
-		}
-
-		newDeploy, err = k8stools.RenderPlaceholders(newDeploy, defaultPlaceholders)
-		if err != nil {
-			return fmt.Errorf("cannot fill placeholders for deployment in vmagent: %w", err)
-		}
-		if err := reconcile.Deployment(ctx, rclient, newDeploy, prevDeploy, false); err != nil {
-			return err
-		}
-		deploymentNames[newDeploy.Name] = struct{}{}
-	case *appsv1.StatefulSet:
-		var prevSTS *appsv1.StatefulSet
-		if prevObjectSpec != nil {
-			prevAppObject, ok := prevObjectSpec.(*appsv1.StatefulSet)
-			if ok {
-				prevSTS = prevAppObject
-				prevSTS, err = k8stools.RenderPlaceholders(prevSTS, defaultPlaceholders)
-				if err != nil {
-					return fmt.Errorf("cannot fill placeholders for prev sts in vmagent: %w", err)
-				}
-			}
-		}
-		newDeploy, err = k8stools.RenderPlaceholders(newDeploy, defaultPlaceholders)
+	if prevCR != nil {
+		prevAppTpl, err = newK8sApp(prevCR, ac)
 		if err != nil {
-			return fmt.Errorf("cannot fill placeholders for sts in vmagent: %w", err)
-		}
-		stsOpts := reconcile.STSOptions{
-			HasClaim:       len(newDeploy.Spec.VolumeClaimTemplates) > 0,
-			SelectorLabels: cr.SelectorLabels,
-		}
-		if err := reconcile.HandleSTSUpdate(ctx, rclient, stsOpts, newDeploy, prevSTS); err != nil {
-			return err
-		}
-		stsNames[newDeploy.Name] = struct{}{}
-	case *appsv1.DaemonSet:
-		var prevDeploy *appsv1.DaemonSet
-		if prevObjectSpec != nil {
-			prevAppObject, ok := prevObjectSpec.(*appsv1.DaemonSet)
-			if ok {
-				prevDeploy = prevAppObject
-			}
+			return fmt.Errorf("cannot build new deploy for vmagent: %w", err)
 		}
-		if err := reconcile.DaemonSet(ctx, rclient, newDeploy, prevDeploy); err != nil {
-			return err
+		if reflect.TypeOf(newAppTpl) != reflect.TypeOf(prevAppTpl) {
```

Comment:
```suggestion
		if prevAppTpl.GetObjectKind() != newAppTpl.GetObjectKind() {
```
Nitpick: this would remove `reflect` import and would probably be bit faster